src/device_memory_tracker.h

```diff
@@ -0,0 +1,265 @@
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
+  }
+}
+
+void
+bufferCompleted(
+    CUcontext ctx, uint32_t streamId, uint8_t* buffer, size_t size,
+    size_t validSize)
+{
+  CUptiResult status;
+  CUpti_Activity* record = nullptr;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(buffer, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+        DeviceMemoryTracker::TrackActivity(record);
+      } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED)
+        break;
+      else {
+        CUPTI_CALL(status);
+      }
+    } while (1);
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
+}
+
+}  // namespace
+
+void
+DeviceMemoryTracker::InitTrace()
+{
+  if (tracker_ == nullptr) {
+    tracker_.reset(new DeviceMemoryTracker());
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_MEMORY2));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+  }
+}
+
+void
+DeviceMemoryTracker::TrackThreadMemoryUsage(ScopedMemoryUsage* usage)
+{
+  if (tracker_ == nullptr) {
+    InitTrace();
+  }
+  CUPTI_CALL(cuptiActivityPushExternalCorrelationId(
+      CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN,
+      reinterpret_cast<uint64_t>(usage)));
+  usage->tracked_ = true;
+}
+
+void
+DeviceMemoryTracker::UntrackThreadMemoryUsage(ScopedMemoryUsage* usage)
+{
+  uint64_t id;
+  if (tracker_ == nullptr) {
+    InitTrace();
+  }
+  CUPTI_CALL(cuptiActivityPopExternalCorrelationId(
+      CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN, &id));
+  CUPTI_CALL(cuptiActivityFlushAll(0));
+  usage->tracked_ = false;
+}
+
+void
+DeviceMemoryTracker::TrackActivityInternal(CUpti_Activity* record)
+{
+  switch (record->kind) {
+    case CUPTI_ACTIVITY_KIND_MEMORY2: {
+      CUpti_ActivityMemory3* memory_record = (CUpti_ActivityMemory3*)record;
+      ScopedMemoryUsage* usage = nullptr;
+      {
+        std::lock_guard<std::mutex> lk(mtx_);
+        auto it = activity_to_memory_usage_.find(memory_record->correlationId);
+        if (it != activity_to_memory_usage_.end()) {
+          usage = reinterpret_cast<ScopedMemoryUsage*>(it->second);
+          activity_to_memory_usage_.erase(it);
+        }
+      }
+      // Igore memory record that is not associated with a ScopedMemoryUsage
+      // object
+      if (usage != nullptr) {
+        const bool is_allocation =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_ALLOCATION);
+        if (is_allocation || (memory_record->memoryOperationType ==
```

Comment:
How about adding another variable `is_release` just for the consistency. It will make the code more readable.